File: Tree/Maximum Depth of Binary Tree.py
# Solution - Recursive
# Time O(N) | Space O(N)

def maxDepth(root):
    if not root:
        return 0
    
    left = maxDepth(root.left)
    right = maxDepth(root.right)
    
    return 1 + max(left, right)

# Solution - Iterative 
# Time O(N) | Space O(N)

def maxDepth(root):
    if not root:
        return 0
    
    stack = [(root, 1)]
    depth = 1
    while stack:
        node, cur = stack.pop()
        depth = max(depth, cur)
        if node.left:
            stack.append((node.left, cur + 1))
        if node.right:
            stack.append((node.right, cur + 1))
            
    return depth
        
        

File: tree-problems/maximumDepthOfBinaryTree.js
function maximumDepth(root)
{
    
    return traverse(root,0)
}

function traverse(root,currentLength)
{
    if(!root)return currentLength

    if(root.left === null )
    {
        return traverse(root.right,currentLength+1)
    }

    if(root.right === null )
    {
        return traverse(root.left,currentLength+1)
    }

    let leftBranch=traverse(root.left,currentLength+1)
    let rightBranch=traverse(root.right,currentLength+1)

    return Math.max(leftBranch,rightBranch)
}